### ISSUE TITLE
helm-help is unusable with Emacspeak

### ISSUE THREAD

**Issue by parhamdoustdar** [emacs-helm/helm#2221]
Currently, it's impossible to use [Emacspeak](https://github.com/tvraman/emacspeak) with the way Helm's `help event loop` is structured.

Emacspeak has advised certain functions to be able to speak. For example, `forward-line` causes an advice to run which reads the current line.

UNfortunately, `read-key` also has a piece of advice tied to it which interrupts speech and reads the `read-key` prompt, which means every time I press a key, because of the `helm-awhile` function being run and quickly using `read-key` to listen for my next key press, I can't read the buffer any more.

Is it possible to perhaps remove the `read-key` and just use the normal Emacs facilities? For example, `helm-help` could just open a help buffer in another window, and maybe put this on the first line:

`"[SPC,C-v,next:ScrollUp  b,M-v,prior:ScrollDown TAB:Cycle M-TAB:All C-s/r:Isearch q:Quit]"`

As I see it, apart from the accessibility issue, the current approach has a few flaws:

- It remaps my existing keys. For example, if I have bound `forward-word` to something else, it will suddenly stop working and I'll need to read the code to know why (in the `read-key` `cl-case`, it expects the key to be `M-f`)
- Any other key that the current `cl-case` doesn't cover will not be usable in this temporary buffer
- The set of keys in the help message that is repeated over and over is actually not complete when looking at the code. Why not make a `helm-help-mode` and use Emacs' own self-documenting awesomeness?

I'm sure there is a valid reason for having the help this way. Please don't take this as a pointless criticism – I'm mainly very excited about Helm and want to be able to use it.

Thanks!

**Comment by thierryvolpiatto** [comment]
Parham Doustdar <notifications@github.com> writes:

> Currently, it's impossible to use Emacspeak with the way Helm's help event loop is structured.
>
> Emacspeak has advised certain functions to be able to speak. For example, forward-line causes an advice to run which reads the current line.
>
> UNfortunately, read-key also has a piece of advice tied to it which
> interrupts speech and reads the read-key prompt,

Isn't this configurable to not read the read-key prompt?

> which means every time I press a key, because of the helm-awhile
> function being run and quickly using read-key to listen for my next
> key press, I can't read the buffer any more.
>
> Is it possible to perhaps remove the read-key and just use the normal
> Emacs facilities?

No, not now at least.  The helm help is really specific to helm, the
goal beeing to have an embeded help without quitting helm.
I could make a rewrite by using a persistent action for help and display
the buffer with special-display-regexps or buf names, but this need a
complete rewrite and will not happen soon.

You can have the whole documentation with M-x helm-documentation though.

> For example, helm-help could just open a help buffer in another
> window, and maybe put this on the first line:
>
> "[SPC,C-v,next:ScrollUp b,M-v,prior:ScrollDown TAB:Cycle M-TAB:All C-s/r:Isearch q:Quit]"
>
> As I see it, apart from the accessibility issue, the current approach has a few flaws:
>
>   • It remaps my existing keys. For example, if I have bound forward-word to something else, it will suddenly stop working and I'll need to read the code to know why
>     (in the read-key cl-case, it expects the key to be M-f)
>   • Any other key that the current cl-case doesn't cover will not be usable in this temporary buffer
>   • The set of keys in the help message that is repeated over and over is actually not complete when looking at the code. Why not make a helm-help-mode and use Emacs'
>     own self-documenting awesomeness?
>
> I'm sure there is a valid reason for having the help this way. Please don't take this as a pointless criticism – I'm mainly very excited about Helm and want to be
> able to use it.
>
> Thanks!
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.*


--
Thierry

Get my Gnupg key:
gpg --keyserver pgp.mit.edu --recv-keys 59F29997

**Comment by thierryvolpiatto** [comment]
Hello and sorry for late reply, but I had hard work to release helm 3.6.0.
Here a first shoot at what we could do to fix this issue, it is an unfinished work not fully tested:

``` elisp
(defun helm-test-special-display-pa (candidate)
  (let ((buf (get-buffer-create "*test*"))
        (source (or (helm-get-current-source)
                    (helm-comp-read
                     "Help for: "
                     (cl-loop for src in (with-helm-buffer helm-sources)
                              collect `(,(assoc-default 'name src) .
                                        ,src))
                     :allow-nest t
                     :exec-when-only-one t)))
        (inhibit-read-only t))
    (with-current-buffer buf
      (erase-buffer)
      (save-excursion
        (helm-aif (assoc-default 'help-message source)
            (insert (substitute-command-keys
                     (helm-interpret-value it)))
          (insert "* No specific help for this source available."))
        (insert "\n\n"
                (substitute-command-keys
                 (helm-interpret-value helm-help-message))))
      (org-mode)
      (org-mark-ring-push)
      (setq buffer-read-only t)
      (local-set-key (kbd "q") 'delete-frame))
    (display-buffer buf '(display-buffer-pop-up-frame))))

(defun helm-run-special-display-help ()
  (interactive)
  (with-helm-alive-p
    (helm-attrset 'help-action 'helm-test-special-display-pa)
    (helm-execute-persistent-action 'help-action)))
(define-key helm-map (kbd "C-c ?") 'helm-run-special-display-help)
```

**Comment by thierryvolpiatto** [comment]
Two problems with the code above:
- Doesn't work in non graphic display.
- Doesn't work with empty sources.
May be other problems depending what the user do in the help frame e.g. run other helm commands from there.
So what I did is I have added two new hooks to helm-help that will allow you disabling the read-key advice, hooks are `helm-help-mode-before-hook` and `helm-help-mode-after-hook`.
Let me know if that's enough.
